### DISCUSSION TITLE
Is Docker secrets actually used in production scenarios?

### DISCUSSION THREAD

**Discussion by jmdfm** [BretFisher/ama#128] [Q&A]
My production app requires around 30 environment variables to be set currently, passing in all sorts of secrets.

Are we really expected to create 30 little files in this case, and have them lying around locally? And what about the case of sharing the secrets between developers? This seems like a really clumsy and error prone way of handling secrets. I can't for the life of me figure out why Docker doesn't have a more realistic and useful way of handling bulk secrets, like say one encrypted YAML file holding all secrets, and have that be the source for creating production secrets for the swarm.

Thanks!

**Comment by BretFisher** [comment]
Correct, you would no do that in that case. Like many things with Docker there is more then one way to solve the problem. That lecture on full lifecycle (I assume it's the one you're referencing) just showed one way.

In your case, if you had enough complexity, you'd likely end up with multiple dockerfiles. One docker-compose.yml for development and maybe automated testing, and a stack.yml for production. You might even have a docker-compose.yml for local dev, a [docker-compose.override.yml for individual dev's custom local settings](https://docs.docker.com/compose/extends/) (not saved in git), and then a custom override just for testing CI/CD systems. The stack.yml would again, be just production.

**Environment Vars:**

You said environment variables *and* secrets, so I hope you realize your env's have lots of options for how to capture them in the compose files, and they're easier to deal with then secrets for unencrypted needs. Obviously, you can use ENV in dockerfiles, that can be overwritten by compose file "environment" per service. You can also do template vars in the compose file itself, the compose format [page has good info at the bottom about](https://docs.docker.com/compose/compose-file/#variable-substitution). It can even now do [YAML extensions to have reusable blocks of data](https://docs.docker.com/compose/compose-file/#extension-fields) (including repetitive env vars) at the top of the file and keep everything [D.R.Y.](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) Also, compose will pickup env vars set in the shell or in a [.env file at the time of running it](https://docs.docker.com/compose/environment-variables/#the-env-file). Whew.

**Secrets:**

The purpose of using docker-compose with secrets files is to emulate what Swarm does with secrets by passing each one in as a file to the container. Note that what local dev does with docker-compose CLI isn't truly secret. It's not using the same methods as Swarm does (which encrypts on host disk, never stored on disk inside container, etc.). We needed a way to develop locally in a way that will "feel like the same result inside the container" but without the requirement of a encrypted store just for dev/test setups.

The local secrets for dev should be dummy passwords when possible. They are not meant to be used in Swarm/prod setups like that (because in swarm the secrets will be in the Raft db already 90% of time, put in by devops person). For the teams I work with we put dummy passwords in a repo subdir like ./sample-secrets or a test subdir that's not stored in image.

In the rare case that you need to handle *real* secrets for local dev (ssh keys, etc.) then docker isn't going to handle the security part of that for you... you'll need to figure out what's acceptable in your team for distributing/storing the team's secrets for dev (like maybe a API key just used for testing but you don't want it stored in repo).

So yea, in dev, if you need to simulate Swarm secrets for local dev you'd have 30 files in a subdir, all controlled via docker-compose.yml.

Now once you get to production, you'll likely be storing secrets in larger teams with something like Vault or AWS Secrets Manager, etc.  You'll wire up automation for those secrets to get injected into docker secret commands... that will match your Swarm stack file names/aliases when you deploy stacks.

**Containers using secrets:**

Now you'll want to emulate how it will work in production for your apps, where they get their secrets via a file, and not env vars like we all used to do. This will ensure your apps work as expected with single files per secret inside the container. Docker's official images like show how to handle these secrets during container startup by using ENTRYPOINT scripts to manage mysql settings for example (or postgres, etc). Many existing apps will need a "helper script" in containers to get file-based secrets into the apps that need them.

Really, there's just so much there I could do an entire course on just docker-compose file features :). 

Feel free to ask further questions here. It's a complex topic.

**Comment by BretFisher** [comment]
More thoughts: 

The "secrets in containers as individual files" is definitely becoming one of the two main standards for apps getting their secrets **in dynamic container-based clusters.** Vault and AWS Secrets Manager work in a different way. In Swarm you wouldn't *replace* Swarm Secrets with Vault, but rather in a large team with complex and very dynamic secret workflows, they would setup Swarm Secrets (which knows how to inject the proper secret into the proper container) so it pulled secrets from Vault... but I don't know anyone doing this yet, likely because it's overkill for small teams and adds complexity without clear benefits (unless you have other Vault needs outside of Swarm and want to do the work of automating the two, but most peoples secrets don't change so often that a few one-line commands won't fix).

The frustration of not having a clear workflow is because there is no "right way". There isn't even a "default way" because it all depends on a number of factors and needs that are unique to you.

A different conversion/question you might ask is something like (for example):

===

> I have a node.js app with a PHP API and mysql backend db with redis in the middle. I store my code in github, I'm a solo developer with no devops or sysadmin to help me. I plan to host this solution on AWS. I don't want to use their existing products like ECS, EKS, Beanstalk, etc. but would rather create a simple Docker Swarm on a minimum number of redundant nodes.  I don't have a Linux distribution preference and I don't have an existing CI/CD solution for automation. What would be the fastest way to hook all this together for mostly (or fully) automated testing and deployments of app updates, and the least amount of cost and overhead for me?

===

That description helps make decisions on what you need end-to-end, and it's something I try to help educate people on, but not something I can prescribe as a "do this exactly every time".  If you just changed one thing, like "I want to use Beanstalk"... it would change most of the solution (you wouldn't use Swarm in that case).

Sorry I don't have an easy answer for you, but I do DockerCon talks on these topics to try and help people like yourself make the best decisions on tools and workflow up front. Here's the recordings:

[DockerCon Part 1: Taking Docker to Production DockerCon EU 2017](https://dockercon.docker.com/watch/WdAeLaLuSCNQwEp61YVXUt)

[DockerCon Part 2: Building Your Swarm Production Stack DockerCon US 2018](https://dockercon2018.hubs.vidyard.com/watch/k3Cv676wmxAwYDxbvcgcgC)
